src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -2099,7 +2110,8 @@ public override void BeforeTransactionCompletion(ITransaction tx)
 			using (new SessionIdLoggingContext(SessionId))
 			{
 				log.Debug("before transaction completion");
-				FlushBeforeTransactionCompletion();
+				if (tx != null || TransactionContext?.CanFlushOnSystemTransactionCompleted != false)
+					FlushBeforeTransactionCompletion();
```

Comment:
If we are completing transaction from system transaction (`tx` null), check if it supports flushing from there. the `tx != null` may look superfluous, but not completely, since it was even told sometimes to use `ITransaction` inside scopes. In such case the `ITransction.Commit` would not trigger flush without that `tx != null` test.